File: ModoMercado/Core/Domain/Errors/AppError.swift
import Foundation

enum AppError: LocalizedError {
    // MARK: - Validação
    case campoObrigatorioAusente(campo: String)
    case valorInvalido(campo: String, motivo: String)
    case produtoSemPreco(nomeProduto: String)

    // MARK: - Persistência
    case falhaAoSalvar(entidade: String, causa: Error)
    case falhaAoLer(entidade: String, causa: Error)
    case falhaAoExcluir(entidade: String, causa: Error)
    case dadosCorrempidos

    // MARK: - Negócio
    case produtoPossuiRegistros(nomeProduto: String)
    case listaVazia

    // MARK: - LocalizedError

    var errorDescription: String? {
        switch self {
        case .campoObrigatorioAusente(let campo):
            return "O campo '\(campo)' é obrigatório e não pode estar vazio."
        case .valorInvalido(let campo, let motivo):
            return "O campo '\(campo)' possui um valor inválido: \(motivo)."
        case .produtoSemPreco(let nomeProduto):
            return "O produto '\(nomeProduto)' não pode ser adicionado pois não possui nenhum preço registrado."
        case .falhaAoSalvar(let entidade, let causa):
            return "Falha ao salvar \(entidade): \(causa.localizedDescription)"
        case .falhaAoLer(let entidade, let causa):
            return "Falha ao ler \(entidade): \(causa.localizedDescription)"
        case .falhaAoExcluir(let entidade, let causa):
            return "Falha ao excluir \(entidade): \(causa.localizedDescription)"
        case .dadosCorrempidos:
            return "Os dados do aplicativo estão corrompidos e não puderam ser carregados."
        case .produtoPossuiRegistros(let nomeProduto):
            return "O produto '\(nomeProduto)' possui registros de preço vinculados. Ao excluí-lo, todos os registros também serão removidos."
        case .listaVazia:
            return "A lista de compras está vazia."
        }
    }

    var recoverySuggestion: String? {
        switch self {
        case .campoObrigatorioAusente:
            return "Preencha o campo indicado antes de continuar."
        case .valorInvalido:
            return "Corrija o valor do campo indicado e tente novamente."
        case .produtoSemPreco:
            return "Registre ao menos um preço para este produto antes de adicioná-lo à lista."
        case .falhaAoSalvar, .falhaAoLer, .falhaAoExcluir:
            return "Tente novamente. Se o problema persistir, reinicie o aplicativo."
        case .dadosCorrempidos:
            return "O aplicativo será iniciado com um estado vazio. Seus dados anteriores não puderam ser recuperados."
        case .produtoPossuiRegistros:
            return "Confirme a exclusão para remover o produto e todos os seus registros vinculados."
        case .listaVazia:
            return "Adicione itens à lista antes de continuar."
        }
    }
}

File: modo_mercado/lib/core/domain/errors/app_error.dart
/// Hierarquia de erros do Modo Mercado.
sealed class AppError implements Exception {
  const AppError();

  String get mensagem;
  String? get sugestao => null;

  @override
  String toString() => 'AppError: $mensagem';
}

// --- Validação ---

class CampoObrigatorioAusenteError extends AppError {
  final String campo;
  const CampoObrigatorioAusenteError(this.campo);

  @override
  String get mensagem => 'O campo "$campo" é obrigatório.';

  @override
  String? get sugestao => 'Preencha o campo "$campo" antes de continuar.';
}

class ValorInvalidoError extends AppError {
  final String campo;
  final String motivo;
  const ValorInvalidoError(this.campo, this.motivo);

  @override
  String get mensagem => 'Valor inválido no campo "$campo": $motivo.';
}

class ProdutoSemPrecoError extends AppError {
  final String nomeProduto;
  const ProdutoSemPrecoError(this.nomeProduto);

  @override
  String get mensagem =>
      'O produto "$nomeProduto" não pode ser adicionado pois não possui preço registrado.';

  @override
  String? get sugestao => 'Registre um preço para "$nomeProduto" antes de adicioná-lo à lista.';
}

// --- Persistência ---

class FalhaAoSalvarError extends AppError {
  final String entidade;
  final Object causa;
  const FalhaAoSalvarError(this.entidade, this.causa);

  @override
  String get mensagem => 'Falha ao salvar $entidade.';

  @override
  String? get sugestao => 'Tente novamente. Se o problema persistir, reinicie o aplicativo.';
}

class FalhaAoLerError extends AppError {
  final String entidade;
  final Object causa;
  const FalhaAoLerError(this.entidade, this.causa);

  @override
  String get mensagem => 'Falha ao carregar $entidade.';
}

class FalhaAoExcluirError extends AppError {
  final String entidade;
  final Object causa;
  const FalhaAoExcluirError(this.entidade, this.causa);

  @override
  String get mensagem => 'Falha ao excluir $entidade.';
}

class DadosCorrempidosError extends AppError {
  const DadosCorrempidosError();

  @override
  String get mensagem => 'Os dados do aplicativo estão corrompidos.';

  @override
  String? get sugestao => 'O aplicativo será iniciado com dados vazios.';
}

// --- Negócio ---

class ProdutoPossuiRegistrosError extends AppError {
  final String nomeProduto;
  const ProdutoPossuiRegistrosError(this.nomeProduto);

  @override
  String get mensagem =>
      'O produto "$nomeProduto" possui registros de preço vinculados.';

  @override
  String? get sugestao =>
      'Ao excluir o produto, todos os registros vinculados também serão removidos.';
}

class ListaVaziaError extends AppError {
  const ListaVaziaError();

  @override
  String get mensagem => 'A lista de compras está vazia.';
}
